Refactoring/record view model

withseon opened the pull request:
## 개요 (이슈 번호 및 요약)
- #142 
    - RecordViewModel 리팩토링
- #152 
    - RecordViewModel 관련 View 코드 수정

## 변경 사항 (작업 내용)
- ### Drink 모델 오타 수정
    - GenderPreference -> genderPreference
    - DrinkField 모델 drinkImageURL 옵셔널 타입 변경
- ### WrittenUser 모델 변경
    - userAge, userGender 추가
- ### 기존 RecordViewModel -> ExRecordViewModel 변경
- ### RecordViewModel 파일에 코드 새로 작성
    - uploadMultipleImagesToFirebaseStorageAsync() 수정
    - 기존 updateDrinkDataWithTag() -> updateDrinkToFirestore() 수정
    - 기존 uploadPost() -> uploadPostToFirestroe() 수정
    - 기존 updatePost() -> updatePostToFirestore() 수정 
- ### FireStorageService 내 uploadImageToStorage() 코드 수정
     ```swift
     // user profile image 업로드 시, 경로 userImages/fileName
     // post images 업로드 시, 경로 postsImages/userID/postID/fileName
     func uploadImageToStorage(folder: FireStorageFolderType, userID: String? = nil, postID: String? = nil, image: UIImage, fileName: String) async throws {
        do {
            var storagePath = folder.rawValue
            if let userID = userID, let postID = postID {
                storagePath += "/\(userID)/\(postID)"
            }
            let storageRoute = storageRef.child("\(storagePath)/\(fileName)")
            let data = Formatter.compressImage(image)
            let metaData = StorageMetadata()
            metaData.contentType = imageType
            guard let data = data else { return }
            let _ = try await storageRoute.putDataAsync(data, metadata: metaData)
        } catch {
            throw FireStorageError.uploadImage
        }
    }
    ```
- ### FirestorePostService 수정
    - FireStorage에 post images를 올릴 때, 폴더링을 위해 postID가 필요함
    - 기존 uploadPostDocument에서 uuid를 부여하는 코드를 제거하고, RecordViewModel에서 postID를 가지고 있도록 수정
    ```swift
    func uploadPostDocument(post: Post, postID: String) async throws {
        ...
    }
    ``` 
- ### RecordViewModel 관련 View 코드 수정
   - AddTagView
   - DrinkTagScroll
   - PhotoSelectPagingTab
   - SearchTagView
   - RecordView
       - 함수 RecordViewModel 이동
       - recordType에 따른 분기 처리
   - SelectedPhotoHorizontalScroll
       - recordType에 따른 분기 처리
           - add인 경우: 기존 코드
           - eidt인 경우: 
           ```swift
           if let post = recordViewModel.post {
               // post가 가진 ImagesURL을 KFImage로 보여주기
               ForEach(0..<(post.postField.imagesURL.count), id: \.self) { index in
                   SelectedImageKFImage(url: post.postField.imagesURL[index])
               }
           }
           ```
- ### Review comment 내용 반영
    - RecordViewModel 내 변수명 변경 및 다른 View 적용

## 스크린샷

## 특이사항 (트러블 슈팅 과정 및 참고 자료 등)

editPost 메소드 위치에 대해
- PostViewModel -> RecordViewModel 이동은 어떨까요..?
- 글 수정 시 RecordView를 재사용하여 수정작업이 이루어지기 때문에 RecordViewModel이 더 적합하지 않을까 하는 생각이 듭니다.

withseon committed:
[Edit] RecordViewModel 리팩토링 #142

- WrittenUser 모델 변경
- Drink 모델 오타 수정
- 기존 RecordViewModel -> ExRecordViewModel 변경
- RecordVieModel 새로 생성 및 코드 리팩토링
- FireStroageService 의 uploadImageToStorage() 수정

withseon committed:
[Edit] error message 추가 #142

withseon committed:
[Edit] 코드 일부 수정

Phangg commented on JUDA/ViewModel/Record/RecordViewModel.swift:
```diff
@@ -2,219 +2,157 @@
-final class RecordViewModel: ObservableObject { 
-    // post 업로드용 Post 모델 객체
+@MainActor
+final class RecordViewModel {
```
user 보다는 writtenUser,
images 보다는 selectedImages 가 조금 더 뭔가 확 이해가 될 것 같은데 어때요..?
의견입니다..

Phangg replied:
고생하셨어요!! 👍

bdrsky2010 commented:
확인완료

withseon committed:
[Edit] RecordViewModel 관련 View 코드 수정 (중간 커밋) #152

- DrinkField 데이터 모델 drinkImageURL 옵셔널 타입 변경
- FirestoreDrinkService 내 deleteDrinkLikedUsersID() 추가
- DrinkViewModel 내 addOrRemoveToLikedUsersID() 추가
- RecordViewModel 일부 수정
- AddTagView, DrinkTagScroll, SearchTagView, RecordView 코드 수정

withseon committed:
[Edit] RecordViewModel 관련 View 코드 수정 #152

- DrinkViewModel, FirestoreDrinkService 함수 중복으로 인한 삭제
- RecordView, SelectedPhotoHorizontalScroll recordType에 따른 분기 처리
- 그 외 코드 수정

Phangg commented on JUDA/View/Record/AddTagView.swift:
```diff
@@ -11,7 +11,7 @@ import PhotosUI
 struct AddTagView: View {
     @EnvironmentObject private var navigationRouter: NavigationRouter
-    @EnvironmentObject private var auth: AuthService
+//    @EnvironmentObject private var auth: AuthService
```
안쓰는건 지워도 될 것 같아요! ㅎㅎ

Phangg commented on JUDA/View/Record/AddTagView.swift:
```diff
@@ -11,7 +11,7 @@ import PhotosUI
 struct AddTagView: View {
     @EnvironmentObject private var navigationRouter: NavigationRouter
-    @EnvironmentObject private var auth: AuthService
+//    @EnvironmentObject private var auth: AuthService
```
안쓰는건 지워도 될 것 같아요! ㅎ

withseon replied:
주석 삭제하는 걸 깜빡했네요.
수정 완료했습니다 -!

Phangg commented on JUDA/View/Record/RecordView.swift:
```diff
@@ -103,72 +110,50 @@ struct RecordView: View {
                     }
-					DispatchQueue.main.async {
-						Task {
-							await postUploadButtonAction()
```
기존에 `postUploadButtonAction` 메서드가 ViewModel 로 이동한 것은 확인했는데, 여기서 호출하는 코드가 없네요..?
다른쪽에서 호출 된건가요?

withseon replied:
기존 Firestore에 post를 업로드하는 postUpload() 함수를 postUploadToFirestore()로 변경하고
게시글 작성 완료 시 호출되는 postUploadButtonAction()을 postUpload()로 이름 변경 후
내부에서 FireStorage의 이미지 작업과 Firestore의 post 업로드, drink를 업데이트하도록 했습니다.

Phangg replied:
제가 꼼꼼히 보지 못했던 것 같아요! 설명 갑사합니다 👍

Phangg commented on JUDA/View/Record/SearchTagView.swift:
```diff
@@ -64,7 +66,7 @@ struct SearchTagView: View {
                     }
                 // 검색 중
-                } else if searchDrinkViewModel.isLoading {
+                } else if drinkViewModel.isLoading {
```
이거 TODO 좀 달아주세요!
제가 isSearching 을 하나 만들었는데, 여기서 사용 될 것 같아요!
나중에 merge 하고 수정 같이 하면 좋을 것 같아요 ㅎㅎ

withseon replied:
넵 수정했습니다 -!

Phangg commented on JUDA/View/Record/SelectedPhotoHorizontalScroll.swift:
```diff
@@ -28,3 +44,25 @@ struct SelectedPhotoHorizontalScroll: View {
 }
+
+// MARK: - 술 리스트 셀 사용 KingFisher 이미지
+struct SelectedImageKFImage: View {
```
이거 저 나중에 한번 싹- 수정하려고 하긴 하는데 미리 수정해도 좋을 것 같아서요! 
딱 현재 뷰에서만 사용되는 뷰 (구조체 ) 이기 때문에,
private struct 로 해놓는게 좋을 것 같아요 ㅎㅎ

withseon replied:
접근제어자는 항상 어렵네요 ㅠㅠ
수정완료 했습니다 !

Phangg commented on JUDA/ViewModel/Drink/FirestoreDrinkService.swift:
```diff
@@ -196,6 +196,26 @@ extension FirestoreDrinkService {
 	}
 }
 
+extension FirestoreDrinkService {
```
요건 어디서 쓰이나요??
좋아요 눌렀을 때, 바꿔주는 건가요? CloudFunction 에서 해주는 건가 해서요..!

withseon replied:
post 업로드 했을 때, 특정 술의 점수가 4점 이상인 경우
root drink의 agePreference, genderPreference값을 변경해주기 위해 작성했습니다.
root 외의 drink 수정 사항은 Cloud Function에서 처리될 예정입니다..!

Phangg replied:
아하- 그 디테일이.. 🫡 저는 생각도 못했네요ㅜ

Phangg commented on JUDA/ViewModel/Record/RecordViewModel.swift:
```diff
@@ -2,219 +2,244 @@
+    // post 작성자 정보를 갖는 WrittenUser 객체
+    @Published var writtenUser: WrittenUser?
+    // post의 전체 술 평가 정보를 갖는 DrinkTag 배열
+    @Published var drinkTags = [DrinkTag]()
```
post 의 타입이 Post 라서, 
post 를 fetch 받고나면 
post.user 로 WrittenUser 에 접근 할 수 있고,
post.drinkTags 로 [DrinkTag] 에 접근할 수 있어서 
필요없는 프로퍼티가 있는 것 같아요

withseon replied:
edit의 경우 fetch된 post로 접근이 가능하지만, 
add의 경우 post 객체를 생성하기 전 유저와 술 태그 정보를 받아둘 프로퍼티가 필요한 거 같아 작성했습니다. :)

Phangg replied:
그렇군요! 그러면 주석에 수정 시 사용한다고 한번 해주시면 알기 좋을 것 같아요! ㅎㅎ
설명 감사합니다 👍

Phangg commented on JUDA/ViewModel/Record/RecordViewModel.swift:
```diff
@@ -2,219 +2,244 @@
-    private let drinkImagesPath = "drinkImages/"
-}
+    // post 업로드, iamges 업로드를 위한 postID
+    private var postID = ""
```
어디에서 사용되고 있나요?

withseon replied:
게시글을 작성할 때,
1. FireStorage에 이미지 업로드 후 URL 받기
2. post 객체 생성
3. Firestore에 post 업로드

순서로 진행되는데, "1. FireStorage에 이미지 업로드 후 URL 받기"에서 postImages/userID/postID/fileName 과 같은 폴더링을 방식을 채택하게 되면서 postID를 UUID로 부여하는 과정이 0순위가 되어버렸습니다.

따라서 게시글 작성 완료 시 호출되는 게시글 작성 완료 시 호출되는 postUpload()(이전 postUploadButtonAction())에서 postID를 UUID로 부여하고, recordPostDataClear()에서 postID를 비워주고 있습니다...

postID가 중요한 정보인데, 여기 저기서 바뀌어지면 안 될 거 같다는 마음이 저도 들었습니다...😢

Phangg commented on JUDA/ViewModel/Record/RecordViewModel.swift:
```diff
@@ -2,219 +2,244 @@
-	}
+    @Published var foodTags = [String]()
+    // 글 작성 or 수정 기준 충족 ( 글 내용 필수 )
+    @Published var isPostContent: Bool = false
```
'기준 충족' 이라는게 인식되는 네이밍이면 좋을 것 같아요!

withseon replied:
'없다'는 걸 파악하는 편이 더 보기 쉽다고 생각하여
isPostContent -> postConistentIsEmpty로 변경하고 기본값을 true로 변경했습니다. 이에 따라 ture/false 변경 부분 코드를 수정하였습니다.
그리고 기존 isPostContentNotEmpty() -> checkpostContentIsEmpty()로 변경하였습니다.

Phangg commented on JUDA/ViewModel/Record/RecordViewModel.swift:
```diff
@@ -2,219 +2,244 @@
-    var imagesURL: [URL] = []
+    @Published var selectedImages = [UIImage]()
+    // 모든 imagesURL을 갖는 배열
+    var imagesURL = [URL]()
```
어디에서 사용되고 있나요?

withseon replied:
uploadMultipleImagesToFirebaseStorageAsync() 에서 image를 FireStorage에 upload한 후, URL을 받아오기 위한 프로퍼티입니다.
게시글 작성 완료 시 호출되는 postUpload()(이전 postUploadButtonAction()) 내에서 post 객체를 생성할 때, imgesURL에 해당 프로퍼티의 값을 넣어줍니다.

withseon committed:
[Edit] review comment에 따른 코드 수정

bdrsky2010 commented:
확인했습니다. 머지하고 다시 한 번 보겠습니다
